File: modules/detection/detection.go
/*
Copyright ou © ou Copr. Cécile Rolland et Charles Mailley, (21 janvier 2025)

aquarium[@]mailo[.]com

Ce logiciel est un programme informatique servant à l'analyse des collectes
traçologiques effectuées avec le logiciel DFIR-ORC.

Ce logiciel est régi par la licence CeCILL soumise au droit français et
respectant les principes de diffusion des logiciels libres. Vous pouvez
utiliser, modifier et/ou redistribuer ce programme sous les conditions
de la licence CeCILL telle que diffusée par le CEA, le CNRS et l'INRIA
sur le site "http://www.cecill.info".

En contrepartie de l'accessibilité au code source et des droits de copie,
de modification et de redistribution accordés par cette licence, il n'est
offert aux utilisateurs qu'une garantie limitée.  Pour les mêmes raisons,
seule une responsabilité restreinte pèse sur l'auteur du programme,  le
titulaire des droits patrimoniaux et les concédants successifs.

A cet égard  l'attention de l'utilisateur est attirée sur les risques
associés au chargement,  à l'utilisation,  à la modification et/ou au
développement et à la reproduction du logiciel par l'utilisateur étant
donné sa spécificité de logiciel libre, qui peut le rendre complexe à
manipuler et qui le réserve donc à des développeurs et des professionnels
avertis possédant  des  connaissances  informatiques approfondies.  Les
utilisateurs sont donc invités à charger  et  tester  l'adéquation  du
logiciel à leurs besoins dans des conditions permettant d'assurer la
sécurité de leurs systèmes et ou de leurs données et, plus généralement,
à l'utiliser et l'exploiter dans les mêmes conditions de sécurité.

Le fait que vous puissiez accéder à cet en-tête signifie que vous avez
pris connaissance de la licence CeCILL, et que vous en avez accepté les
termes.
*/

package detection

import (
	"aquarium/modules/aquabase"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"strings"
	"time"
)

/* VARIABLES LOCALES */
var cheminRegles string = ""

type Regle struct {
	Nom         string    `json:"nom"`
	Auteur      string    `json:"auteur"`
	Description string    `json:"description"`
	Criticite   int       `json:"criticite"`
	Date        time.Time `json:"date"`
	SQL         string    `json:"sql"`
	IsGlobal    bool
}

type regleSQL struct {
	Nom string `json:"nom"`
	SQL string `json:"sql"`
}

/* FONCTIONS LOCALES */

func lancerRegle(cheminProjet string, cheminRegle string) (int, error) {
	// Recuperation de la requete SQL associée à la règle
	var detailsRegle regleSQL
	donneesFichier, err := os.ReadFile(cheminRegle)
	if err != nil {
		log.Println("WARN | Le fichier de règle "+cheminRegle+" n'existe pas ou n'a pas pu être ouvert : ", err.Error())
		return 0, err
	}
	err = json.Unmarshal(donneesFichier, &detailsRegle)
	if err != nil {
		return 0, err
	}
	ruleName := strings.Replace(filepath.Base(cheminRegle), ".json", "", -1)

	// Execution la requête SQL
	var adb = aquabase.InitDB_Extraction(cheminProjet)
	df := adb.SelectFrom0(detailsRegle.SQL)
	isError := df.Table.Nrow() > 0

	// Renseignement de la table sql des regles
	adb_rules := aquabase.InitDB_Rules(cheminProjet)
	tableName := "regles"
	colName := []string{"name", "isError"}
	tableColumns := map[string]string{
		// "id" en autoincrement par defaut
		"name":    "TEXT UNIQUE",
		"isError": "INTEGER",
	}
	err = adb_rules.CreateTableIfNotExist2(tableName, tableColumns, true)
	if err != nil {
		return 0, err
	}
	err = adb_rules.InsertOrReplace(tableName, colName, []interface{}{ruleName, isError})
	if err != nil {
		return 0, err
	}
	//fmt.Println(adb_rules.SelectFrom0("SELECT * FROM regles"))

	// Renvoi 2 si le dataframe n'est pas vide sinon 1
	if isError {
		id_frame := adb_rules.SelectFrom0("SELECT id FROM regles WHERE name='" + ruleName + "'")
		id_value := id_frame.Strloc(0, 0)

		table_name := "error_" + id_value
		err := adb_rules.DropTable(table_name)
		if err != nil {
			return 0, err
		}
		err = adb_rules.CreateTableIfNotExist1(table_name, df.Table.Names(), false)
		if err != nil {
			return 0, err
		}
		err = adb_rules.SaveDf(df.Table, table_name)
		if err != nil {
			return 0, err
		}
		return 2, nil
	}
	return 1, nil
}

func emplacementRegles() string {
	if cheminRegles != "" {
		return cheminRegles
	} else {
		emplacementExecutable, err := os.Executable()
		if err != nil {
			return ""
		}
		emplacementExecutable, err = filepath.EvalSymlinks(emplacementExecutable)
		if err != nil {
			return ""
		}
		// On cherche la liste des règles
		cheminRegles = filepath.Join(filepath.Dir(emplacementExecutable), "ressources", "regles_detection")
		return cheminRegles
	}
}

/** Extract all rule names from a folder
 * @rulesPath : the path folder where rules.json exists
 * @return : list of name rules
 */
func searchDetectionRules(rulesPath string, args ...string) ([]string, error) {
	var listeRegles []string
	// Vérifier si une clé est fournie
	var key string
	if len(args) > 0 {
		key = args[0]
	}
	// Open in kernel the folder
	fichiersRegles, err := os.ReadDir(rulesPath)
	if err != nil {
		return nil, err
	}
	// Catch all json file in a list
	for _, fichierRegle := range fichiersRegles {
		if strings.HasSuffix(fichierRegle.Name(), ".json") {
			nomRegle := strings.TrimSuffix(fichierRegle.Name(), ".json")
			if key == "" || strings.EqualFold(nomRegle, key) {
				listeRegles = append(listeRegles, nomRegle)
			}
		}
	}
	return listeRegles, nil
}

/*
*
return the path of the rule if it exist and if its a global rule
*/
func getRulePathFolder(cheminProjet string, nomRegle string) (string, bool) {
	// Search where the rule is saved in local
	path_local := filepath.Join(cheminProjet, "regles_detection")
	existInLocal, _ := searchDetectionRules(path_local, nomRegle)
	if len(existInLocal) == 1 {
		return path_local, false
	}
	// Search where the rule is saved in global
	path_global := emplacementRegles()
	existInGlobal, _ := searchDetectionRules(path_global, nomRegle)
	if len(existInGlobal) == 1 {
		return path_global, true
	}
	// The rule doesnt exist
	return "", false
}

/********************************************************************************/
/****************************** FONCTIONS GLOBALES ******************************/
/********************************************************************************/

/* Fonction qui renvoie une liste de règles associées à leur état (0:non lancé, 1:négatif, 2:positif) */
func ListeReglesDetection(cheminProjet string, lancerRegles bool) (map[string]map[string]int, []string, error) {
	listeRegles := make(map[string]map[string]int)
	// Search all rule files
	path_local := filepath.Join(cheminProjet, "regles_detection")
	path_global := emplacementRegles()
	listeRegles_local, error_local := searchDetectionRules(path_local)
	listeRegles_global, error_global := searchDetectionRules(path_global)
	if error_local != nil {
		return nil, nil, error_local
	}
	if error_global != nil {
		return nil, nil, error_global
	}
	// Helper function to handle the rule logic
	var probleme error = nil
	var reglesEnErreur []string = []string{}
	handleRule := func(rule string, isGlobal int, path string) {
		state := 0
		var err error
		if lancerRegles {
			path_rule := filepath.Join(path, rule+".json")
			state, err = lancerRegle(cheminProjet, path_rule)
			if err != nil {
				state = 0
				probleme = err
				log.Println("detection.go => lancerRegle(", rule, ") : ", err)
			} else if state == 0 {
				reglesEnErreur = append(reglesEnErreur, rule)
			}
		} else {
			// Cherche si la regle a deja été executé avant
			adb_rules := aquabase.InitDB_Rules(cheminProjet)
			query := fmt.Sprintf("SELECT isError FROM regles WHERE name=\"%s\"", rule)
			df := adb_rules.SelectFrom0(query)
			if df.Table.Nrow() > 0 {
				value, _ := df.Intloc(0, 0)
				if value == 1 {
					state = 2
				} else {
					state = 1
				}
			}
		}
		listeRegles[rule] = map[string]int{
			"isGlobal": isGlobal,
			"state":    state,
		}
	}
	// Merge both list in a list of dict with parameters of each rule
	// Une regle créé par l'user et prioritaire par rapport à une regle de base
	for _, rule := range listeRegles_global {
		handleRule(rule, 1, path_global)
	}
	for _, rule := range listeRegles_local {
		handleRule(rule, 0, path_local)
	}
	return listeRegles, reglesEnErreur, probleme
}

func DetailsRegleDetection(cheminProjet string, nomRegle string) (Regle, error) {
	var donneesRegle Regle
	// Search where the rule is saved
	path, isGlobal := getRulePathFolder(cheminProjet, nomRegle)
	donneesRegle.IsGlobal = isGlobal
	// Read and extract the rule.json data
	donneesFichier, err := os.ReadFile(filepath.Join(path, nomRegle+".json"))
	if err != nil {
		log.Println("WARN DetailsRegleDetection() | Le fichier de règle n'existe pas ou n'a pas pu être ouvert : ", err.Error())
		return Regle{}, err
	}
	err = json.Unmarshal(donneesFichier, &donneesRegle)
	return donneesRegle, err
}

func ResultatRegleDetection(cheminProjet string, nomRegle string) (int, error) {
	// Search where the rule is saved
	path, _ := getRulePathFolder(cheminProjet, nomRegle)
	// Execute the SQL request
	return lancerRegle(cheminProjet, filepath.Join(path, nomRegle+".json"))
}

func ResultatSQL(cheminProjet string, ruleName string) ([]map[string]interface{}, error) {
	adb_rules := aquabase.InitDB_Rules(cheminProjet)
	id_frame := adb_rules.SelectFrom0("SELECT id FROM regles WHERE name='" + ruleName + "'")
	id_value := id_frame.Strloc(0, 0)
	df := adb_rules.SelectFrom0("SELECT * FROM error_" + id_value)
	return df.ToMap(), nil
}

func SuppressionRegleDetection(cheminProjet string, nomRegle string) error {
	// Search where the rule is saved
	path_local := filepath.Join(cheminProjet, "regles_detection")
	exist, _ := searchDetectionRules(path_local, nomRegle)
	if len(exist) != 1 {
		fmt.Println("Annulation de suppression de la regle: " + nomRegle)
		return nil
	}
	// Delete the rule.json data
	err := os.Remove(filepath.Join(path_local, nomRegle+".json"))
	if err != nil {
		log.Println("WARN | Le fichier de règle n'a pas pu être supprimé : ", err.Error())
		return err
	}
	fmt.Println("Suppression de la regle: " + nomRegle)
	// Delete all data about this rule from regles.db
	adb_rules := aquabase.InitDB_Rules(cheminProjet)
	err = adb_rules.DropTable(nomRegle)
	if err != nil {
		return err
	}
	err = adb_rules.RemoveFromWhere("regles", "name='"+nomRegle+"'")
	if err != nil {
		return err
	}
	return nil
}

func StatutReglesDetection(cheminProjet string) ([]map[string]interface{}, error) {
	adb_rules := aquabase.InitDB_Rules(cheminProjet)
	df := adb_rules.SelectFrom0("SELECT * FROM regles")
	if df.Error != nil {
		fmt.Println("Table 'regle' inexistante ou erreur")
		return []map[string]interface{}{}, nil
	}
	return df.ToMap(), nil
}

func NewDetectionRule(chemin_projet string, json_rule string) error {
	chemin_regles := filepath.Join(chemin_projet, "regles_detection")
	// Conversion de la chaîne JSON en une structure Go
	var regle map[string]interface{}
	if err := json.Unmarshal([]byte(json_rule), &regle); err != nil {
		return err
	}
	// Récupération du nom à partir du JSON
	nom, ok := regle["nom"].(string)
	if !ok {
		return fmt.Errorf("Json without the variable: nom")
	}
	nameBeforeModification, ok := regle["nameBeforeModification"].(string)
	if !ok {
		return fmt.Errorf("Json without the variable: nameBeforeModification")
	}
	//Verification que la regle n'existe pas déjà
	rulePathFolder, _ := getRulePathFolder(chemin_projet, nom)
	if len(rulePathFolder) != 0 && nom != nameBeforeModification {
		return fmt.Errorf("The name '" + nom + "' is already used")
	}
	// Suppression du champ json nameBeforeModification et du json avec l'ancien nom
	if nameBeforeModification != "" {
		err := SuppressionRegleDetection(chemin_projet, nameBeforeModification)
		if err != nil {
			return err
		}
	}
	delete(regle, "nameBeforeModification")
	// Conversion de la structure Go en JSON formaté
	data, err := json.MarshalIndent(regle, "", "  ")
	if err != nil {
		return err
	}
	// Création du chemin complet du fichier avec le nom du JSON
	chemin_complet := filepath.Join(chemin_regles, nom+".json")
	// Écriture des données JSON dans un fichier
	if err := os.WriteFile(chemin_complet, data, 0644); err != nil {
		return err
	}
	return nil
}

File: frontend/src/js/detection.js
/*
Copyright ou © ou Copr. Cécile Rolland et Charles Mailley, (21 janvier 2025) 

aquarium[@]mailo[.]com

Ce logiciel est un programme informatique servant à l'analyse des collectes
traçologiques effectuées avec le logiciel DFIR-ORC. 

Ce logiciel est régi par la licence CeCILL soumise au droit français et
respectant les principes de diffusion des logiciels libres. Vous pouvez
utiliser, modifier et/ou redistribuer ce programme sous les conditions
de la licence CeCILL telle que diffusée par le CEA, le CNRS et l'INRIA 
sur le site "http://www.cecill.info".

En contrepartie de l'accessibilité au code source et des droits de copie,
de modification et de redistribution accordés par cette licence, il n'est
offert aux utilisateurs qu'une garantie limitée.  Pour les mêmes raisons,
seule une responsabilité restreinte pèse sur l'auteur du programme,  le
titulaire des droits patrimoniaux et les concédants successifs.

A cet égard  l'attention de l'utilisateur est attirée sur les risques
associés au chargement,  à l'utilisation,  à la modification et/ou au
développement et à la reproduction du logiciel par l'utilisateur étant 
donné sa spécificité de logiciel libre, qui peut le rendre complexe à 
manipuler et qui le réserve donc à des développeurs et des professionnels
avertis possédant  des  connaissances  informatiques approfondies.  Les
utilisateurs sont donc invités à charger  et  tester  l'adéquation  du
logiciel à leurs besoins dans des conditions permettant d'assurer la
sécurité de leurs systèmes et ou de leurs données et, plus généralement, 
à l'utiliser et l'exploiter dans les mêmes conditions de sécurité. 

Le fait que vous puissiez accéder à cet en-tête signifie que vous avez 
pris connaissance de la licence CeCILL, et que vous en avez accepté les
termes.
*/

/**
 * Execution au chargement de la page
 */
window.onload = function() {
    document.querySelectorAll(".button").forEach(btn => {
        btn.onclick = () => {
            const modal = btn.getAttribute("data-modal");
            document.getElementById(modal).style.display = "block";
        };
    });

    window.onclick = event => {
        if (event.target.className === "modal") {
            event.target.style.display = "none";
        }
    };

    const value = document.querySelector("#criticite_value");
    const input = document.querySelector("#criticite");
    value.textContent = input.value;
    input.addEventListener("input", event => {
        value.textContent = event.target.value;
    });

    // Ajoutez l'événement au bouton pour fermer toutes les règles
    const closeButton = document.querySelector('button[onclick="closeAllRules()"]');
    closeButton.addEventListener("click", closeAllRules);

    afficher_regles(false);
};


/**
 * Recuperation de l'ensemble des regles de detection au format json au chargement de la page
 * @param {boolean} lancer - Si True alors execute la requete SQL de chaque regle
 */
function afficher_regles(lancer, filtres=null) {
    const conteneur = document.getElementById("regles");
    conteneur.innerHTML = "";
    let listRegles;
    parent.window.go.main.App.ListeReglesDetection(lancer).then(resultat => {
        listRegles = new Map(Object.entries(resultat));
        console.log(listRegles);
        listRegles.forEach((values, regle) => {
            // value for html balises
            const contenu_id = regle//.replace(" ", "");
            const contenu_state = values['state'];
            const detail_smiley = (() => {
                switch (contenu_state) {
                    case 0: return " 😴";
                    case 1: return " 🤓";
                    case 2: return " 🥸";
                    default: return "";
                }
            })();
            // html block
            const detail_regle = document.createElement('details');
            detail_regle.innerHTML = `
                    <div id="${contenu_id}" class="contenuDetails" resultat="${contenu_state}"></div>
                    <summary>${regle}<strong class="etatRegle">${detail_smiley}</strong></summary>
                `;
            detail_regle.onclick = () => informations_regle(contenu_id, regle);
            // Add element in the page
            conteneur.appendChild(detail_regle);
        });
        update_summary(filtres)
    });
}

function update_summary(filtres=null) {
    parent.window.go.main.App.ListeReglesDetection(false).then(resul => {
        let nbRules = new Map(Object.entries(resul)).size;
        parent.window.go.main.App.StatutReglesDetection().then(resultat => {
            const errorCount = resultat.filter(item => item.isError === 1).length;
            const nbElement = resultat.length;
            document.getElementById("total").innerHTML = "Total:<br>"+nbRules;
            document.getElementById("notExecuted").innerHTML = "Inconnu:<br>"+(nbRules-nbElement);
            document.getElementById("valided").innerHTML = "Validation:<br>"+(nbElement-errorCount)
            document.getElementById("detected").innerHTML = "Detection:<br>"+errorCount

            // Apply filters if exist
            if (filtres) {
                const conteneur_regles = document.getElementById("regles");
                const regles = conteneur_regles.querySelectorAll('details > div');
                regles.forEach(regle => {
                    const regleName = regle.id;
                    const regleResult = resultat.find(r => r.name === regleName);
                    if (!(
                        (regleResult === undefined && filtres === "notExecuted") ||
                        (regleResult && regleResult.isError === 0 && filtres === "valided") ||
                        (regleResult && regleResult.isError === 1 && filtres === "detected")
                        )) {
                        regle.parentElement.remove(); // Supprime la balise <details> parente
                    }
                });
            }
        });
    });
}

/**
 * Affiche les informations d'une regle au clic
 */
function informations_regle(id, nom_regle) {
    if (document.getElementById(id).childElementCount <= 1) {
        parent.window.go.main.App.InfosRegleDetection(nom_regle).then(resultat => {
            // rule html balise and parameters
            const regle = document.getElementById(id);
            const criticiteColor = ["#18C700", "#72C702", "#C2D16C", "#E3A500", "#F06136", "#D42222"][resultat["criticite"]] || "#000000";
            // html block to add
            const detail_regle_open = `
                <div>
                    <strong>Criticité : </strong>${resultat["criticite"]}
                    <input type="range" min="0" max="5" value="${resultat["criticite"]}" style="accent-color:${criticiteColor}" oninput="this.value=${resultat['criticite']}">
                </div>
                <p class="code sql"><strong>SQL : </strong>${resultat["sql"]}</p>
                <p><strong>Description : </strong>${resultat["description"]}</p>
                <p><strong>Auteur : </strong>${resultat["auteur"]}</p>
            `;
            // Add element in the page
            regle.innerHTML += detail_regle_open;
            // Button to print the dataframe result
            if (regle.getAttribute("resultat") == 0) {
                const bouton = document.createElement("button");
                bouton.className = "bouton_sombre";
                bouton.innerText = "Lancer cette règle";
                bouton.onclick = () => lancer_regle(id, nom_regle);
                regle.appendChild(bouton);
            } else if (regle.getAttribute("resultat") == 2) {
                const bouton = document.createElement("button");
                bouton.className = "bouton_sombre";
                bouton.innerText = "Afficher le résultat";
                bouton.onclick = () => afficher_resultat_regle(id, nom_regle);
                regle.appendChild(bouton);
            }
            // Button for local rules
            if (!resultat["IsGlobal"]) {
                // delete button
                const boutonModif = document.createElement("button");
                boutonModif.onclick = () => modifier_regle_panel(id, nom_regle);
                boutonModif.className = "bouton_sombre";
                boutonModif.innerText = "Modifier";
                boutonModif.style = "background-color:orange; margin-left:1rem;"
                regle.appendChild(boutonModif);
                const boutonDel = document.createElement("button");
                boutonDel.onclick = () => supprimer_regle(id, nom_regle);
                boutonDel.className = "bouton_sombre";
                boutonDel.innerText = "Supprimer";
                boutonDel.style = "background-color:red; margin-left:1rem;"
                regle.appendChild(boutonDel);

            }
            update_summary()
        });
    }
}


/**
 * Lance une sequence d'execution de la requete SQL d'une regle au clic
 */
function lancer_regle(id, nom_regle) {
    parent.window.go.main.App.ResultatRegleDetection(nom_regle).then(resultat => {
        const regle = document.getElementById(id);
        const bouton = regle.querySelector("button");
        const etatRegle = regle.parentNode.querySelector(".etatRegle");
        if (resultat == 1) {
            etatRegle.textContent = " 🤓";
            bouton.remove();
        } else if (resultat == 2) {
            etatRegle.textContent = " 🥸";
            bouton.innerText = "Afficher le résultat";
            bouton.onclick = () => afficher_resultat_regle(id, nom_regle);
        }
        update_summary()
    });
}


/**
 * Ferme tous les volets pliant
 */
function closeAllRules() {
    const detailsElements = document.querySelectorAll("#regles details");
    detailsElements.forEach(detail => {
        detail.removeAttribute("open");
    });
}

/**
 * Verification de la requete SQL du formulaire de creation d'une regle
 */
function validateSQL() {
    const input = document.getElementById("sql");
    const sqlPattern = /SELECT\s.*\sFROM\s.*\sWHERE\s.*/ig;
    input.value = input.value.trim();

    if (sqlPattern.test(input.value)) {
        input.classList.remove('invalid');
        return true;
    } else {
        input.classList.add('invalid');
        alert("La requête SQL doit être au format 'SELECT % FROM % WHERE %': " + input.value);
        return false;
    }
}

/**
 * Creation d'une nouvelle regle
 */
function creation_regle() {
    const regle = {
        "nom": document.getElementById("nom").value,
        "auteur": document.getElementById("auteur").value,
        "description": document.getElementById("description").value,
        "criticite": parseInt(document.getElementById("criticite").value),
        "sql": document.getElementById("sql").value,
        "nameBeforeModification": document.getElementById("nameBeforeModification").value
    };

    const jsonString = JSON.stringify(regle, null, 2);
    parent.window.go.main.App.CreationReglesDetection(jsonString).then(_ => {
        document.getElementById("popup-newRule").style.display = "none";
        document.getElementById("regles").innerHTML = "";
        document.getElementById("nameBeforeModification").value = "";
        afficher_regles(false);
        update_summary()
    });

}

function supprimer_regle(id, nom_regle) {
    parent.window.go.main.App.Delete_rule(nom_regle).then(_ => {
        document.getElementById("regles").innerHTML = "";
        afficher_regles(false);
        update_summary()
    });
}

function modifier_regle_panel(id, nom_regle) {
    parent.window.go.main.App.InfosRegleDetection(nom_regle).then(resultat => {
        document.getElementById("nom").value = nom_regle;
        document.getElementById("auteur").value = resultat["auteur"];
        document.getElementById("description").value = resultat["description"];
        document.getElementById("criticite").value = resultat["criticite"];
        document.getElementById("sql").value = resultat["sql"];
        document.getElementById("nameBeforeModification").value = nom_regle;

        document.getElementById("popup-newRule").style.display = "block";
        update_summary()
    });
}



function afficher_resultat_regle(id, nom_regle) {
     parent.window.go.main.App.ResultatsSQL(nom_regle).then(resultat => {
        if (!resultat) {return;}
        // add results in the popup
        let div_db_infos = document.querySelector("#popup-resultRule #table_values");
        div_db_infos.textContent = ''

        // Créer un conteneur avec une barre de défilement horizontal
        let scrollContainer = document.createElement('div');
        scrollContainer.style.maxHeight = '450px'
        scrollContainer.style.overflow = 'auto';

        creer_tableau_depuis_dico(resultat, scrollContainer)
        div_db_infos.appendChild(scrollContainer);

        document.querySelector("#popup-resultRule .modal-content").style.width = "90%";
        document.getElementById("popup-resultRule").style.display = "block";
        update_summary()
    });
}


